lib/hsa/hsaBeamformUpchan21cm.hpp

```diff
@@ -0,0 +1,94 @@
+/**
+ * @file
+ * @brief Upchannelization of FRB data with 21cm data extracted
+ *  - hsaBeamformUpchan21cm : public hsaCommand
+ */
+
+#ifndef HSA_BEAMFORM_UPCHAN_21CM_H
+#define HSA_BEAMFORM_UPCHAN_21CM_H
+
+#include "hsaCommand.hpp"
+
+/**
+ * @class hsaBeamformUpchan21cm
+ * @brief hsaCommand to upchannelize and downsample FRB data
+ *
+ * This is an hsaCommand that launches the kernel (upchannelize_flip) for
+ * upchannelizing and downsampling FRB data. First, 128 time samples
+ * is traded for 128 upchannelized freq channels via FFT. The freq axis is
+ * "flipped" (or more accurately, rolled) so that the higher power is in
+ * the middle of the bandpass. Then, every 8 freqs and 3 time samples are
+ * averaged in order to downsample to fit the output bandwidth. The 2
+ * polarizations are summed to produce a final output in float of
+ * 1024 beams x 128 times x 16 freq.
+ *
+ * Hyper fine beam data is also output at this stage. Before the FRB data is
+ * downsampled, both polarizations are summed over all time samples for every
+ * frequency. A bandpass filter is applied and the final output is in float of
+ * 1024 beams x 128 freqs x 10 samples
+ *
+ * @requires_kernel    upchannelize_flip.hasco
+ *
+ * @par GPU Memory
+ * @gpu_mem  transposed_output  Input data of size input_frame_len
+ *     @gpu_mem_type            static
+ *     @gpu_mem_format          Array of @c float
+ *     @gpu_mem_metadata        chimeMetadata
+ * @gpu_mem  bf_output          Output data of size output_frame_len
+ *     @gpu_mem_type            staging
+ *     @gpu_mem_format          Array of @c float
+ *     @gpu_mem_metadata        chimeMetadata
+ * @gpu_mem  hfb_output         Output data of size output_hfb_frame_len
+ *     @gpu_mem_type            staging
+ *     @gpu_mem_format          Array of @c float
+ *
+ * @conf   num_elements         Int (default 2048). Number of elements
+ * @conf   samples_per_data_set Int (default 49152). Number of time samples in a data set
+ * @conf   downsample_time      Int (default 3). Downsample factor in time
+ * @conf   downsample_freq      Int (default 8). Downsample factor in freq
+ * @conf   num_frb_total_beams  Int (default 1024). Number of total FRB formed beams
+ * @conf   num_sub_freqs        Int (default 128). Number of sub frequencies
+ *
+ * @todo   Check that the 16 freq axis has the correct orientation
```

Comment:
I really hope this is actually correct at this point :)   Should probably remove the comment.